### DISCUSSION TITLE
Add the Expanded Memory Adapter (XMA)

### DISCUSSION THREAD

**Discussion by Fenix770** [86Box/86Box#4238] [Ideas]
**Is your feature request related to a problem? Please describe.**

No. I saw the latter additions, (PS/2 models 60 and 65sx), this memory adapter (XMA) supports up to 8mb of RAM, and supports 50, 50Z, 55 SX, 60 and 65 SX models. Also has the capability to configure the RAM as LIM EMS 4.0 memory on 286 based models.

I'm not sure if  documentation exists, but the DOS drivers exist. DR-DOS 3.41 includes the EMMXMA.SYS driver for the adapter and PC-DOS 4.0 includes XMA2EMS.SYS driver for this card. (On IBM 386 machines XMAEM.SYS is used to emulate the adapter and XMA2EMS.SYS to provide EMS memory. 

I'm not sure, but I think this card has the capability to provide UMB blocks with QRAM or Move'em. A special version of QEMM for models 50 and 60 also existed, but I could not find it: 

https://books.google.com.ar/books?id=FK9cQfDtQZkC&pg=PT287&lpg=PT287&dq=qemm+for+ps/2+50/60&source=bl&ots=sL4U1o7v8H&sig=ACfU3U3s3NVt93po6IL03Sfh552pGRpv2g&hl=es-419&sa=X&ved=2ahUKEwidkePC3JH5AhX1iJUCHb5KCS0Q6AF6BAgXEAM#v=onepage&q=qemm%20for%20ps%2F2%2050%2F60&f=false

**Describe the solution you'd like**
The implementation of this card some time in the future.

**Additional context**
Here: is the screenshot of the card:
https://ardent-tool.com/memory/IBM_Memory_Expansion.html#1_8MB

**Comment by iamgreaser** [comment]
The MS-DOS 4.00 source code release (which happened today) contains the XMA2EMS and XMAEM drivers, now MIT-licensed. As far as I can tell from the code, there are the following types:

- MXO (or XMO, not sure which name is canon) - up to 2 MB worth of 16 KB pages w/ a 16 MB virtual address space
- XMA 1 - up to 4 MB worth of 4 KB pages w/ a 16 MB virtual address space
- XMA/A - up to 8 MB worth of 4 KB pages w/ a 16 MB virtual address space
- XMA emulator (XMAEM.SYS) - nothing for 86Box to implement AFAIK, but this is how the 386-based PS/2 Model 80 apparently does it

It seems like the cards behave a bit like paged MMUs, but I don't know how they actually work in practice, and I don't have any real hardware to confirm against. They do look like they could be really flexible cards, though - EMS, UMA, HMA seem to be almost obvious in terms of how to make use of them.

**Reply by OBattler** [reply]
The XMA is already on the long-term TODO list, I believe we need some documentation.
